Enable building of fortran bindings

timostrunk opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

timostrunk committed:
Enable building of fortran bindings

timostrunk commented:
@conda-forge-admin, please rerender

timostrunk commented on recipe/build.sh:
```diff
@@ -23,10 +23,10 @@ ${BUILD_PREFIX}/bin/cmake ${CMAKE_ARGS} \
+  -D REQUIRE_CXX_API=ON \
+  -D REQUIRE_CXX_API_COMPILED=OFF \
+  -D ENABLE_FORTRAN=ON \
+  -D ENABLE_PYTHON=OFF \
```
cmake says that it does not know about those variables with LIBINT2 prefix. I removed the prefix according to some old docs from 2016.
When removing them, cmake knows about REQUIRE_CXX_API and ENABLE_FORTRAN.
ENABLE_FORTRAN=ON also generates the fortran module.

ENABLE_PYTHON and REQUIRE_CSS_API_COMPILED are still unknown. I left them here however.

timostrunk replied:
Build of CP2K fails with:
```
Error: Symbol 'libint2_init_2eri' referenced at (1) not found in module 'libint_f'
```
and
```
 ERROR: __LIBINT_MAX_AM is too large
```

timostrunk replied:
The latter error is most probably caused by the toy source. I will let a full build run over night. Do you have insight into the former @loriab ?

timostrunk replied:
Ah just checked the toy build config and eri2 is off. Will try with the full build

loriab replied:
* Yeah, don't worry about the unrecognized variables. I had left the names as the new, regularized cmake names, but the on/off settings were the defaults, so it didn't matter until you enabled Fortran.
* Right, the toy src is enough to build that Hartree--Fock test and not much more.
* Let me know if the full build was enough to get cp2k linked, please. I started the split-packages recipe, then looked at the linux artifact for your latest build and realized that there isn't a fortran library, only the `include/libint_f.mod` file is new.
  From the src, this is expected as only the module file is installed: https://github.com/evaleev/libint/blob/master/export/cmake/CMakeLists.txt.export#L559 . So perhaps this is all cp2k needs? If so, I was wrong about needing split packages and you just need a `# [unix]` selector on the fortran compiler in meta.yaml to "fix" windows. Or maybe more needs to be installed for the fortran build -- the cmake builds an object file but no library.

timostrunk replied:
Yes, the only changed file appears to be the module. This would not be a problem. The larger change is the produced meta.yaml in the fortran build:
```
Without gfortran compiler dep (from your compiled package):
  run:
    - libgcc-ng >=12
    - libstdcxx-ng >=12
With the fortran dep (from my compiled package:
  run:
    - libgcc-ng >=12
    - libgfortran-ng
    - libgfortran5 >=12.3.0
    - libstdcxx-ng >=12
```
In other words: We incur a gfortran dependency for pure C++ users, which is not required for c++ users. The solution could be one build step with two subpackages, where the c++ package uses:
```
build:
  ignore_run_exports_from:
    - {{ compiler('fortran') }}
```
and depends on the c++ package and contains only the module file.

I didn't try the build with CP2K yet, but will soon.

timostrunk replied:
CP2K links correctly. The regtests are still running. I now tried making the fortran build conditional (windows off), but did not split the package yet. Maybe we don't need to considering the fortran lib dependencies are basically everywhere.

timostrunk replied:
Tests passed on the CP2K side, so ENABLE_FORTRAN seems to be the only thing required. I also disabled build for Windows. If we decide to ignore the added fortran library run dependency, this would be good to go, once I remove the two bogus commits and rerender, which I will do now.

timostrunk commented:
@conda-forge-admin, please rerender

loriab commented:
Great news that this works w/o further cmake changes!

* Even though the fortran libraries are usually around by the time one adds BLAS, I don't want to require them for a base libint package. We can split it up into two output packages. Or, because it's only a light header file, not a library, I think we could keep it one output and add the 

```
build:
  ignore_run_exports_from:
    - {{ compiler('fortran') }}
```

to the whole package. It's not quite The Right Way to do things, but the analysis isn't even picking up on the mod file anyways:

```
Warning: rpath /home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/lib is outside prefix /home/conda/feedstock_root/build_artifacts/libint_1704194798938/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac (removing it)
   INFO: sysroot: '/home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/x86_64-conda-linux-gnu/sysroot/' files: '['/home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/zone.tab', '/home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/tzdata.zi', '/home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/right/Zulu', '/home/conda/feedstock_root/build_artifacts/libint_1704194798938/_build_env/x86_64-conda-linux-gnu/sysroot/usr/share/zoneinfo/right/WET']'
   INFO (libint,lib/libint2.so): Needed DSO lib/libstdc++.so.6 found in conda-forge/linux-64::libstdcxx-ng==13.2.0=h7e041cc_3
   INFO (libint,lib/libint2.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libpthread.so.0 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.12=he073ed8_16
   INFO (libint,lib/libint2.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.12=he073ed8_16
WARNING (libint): dso library package conda-forge/linux-64::libgfortran5==13.2.0=ha4646dd_3 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

* Also, the `script_env` isn't necessary b/c the Windows build is using `bld.bat` and so still has Fortran off. (You can correct the cmake variable names if you want.)

How does this sound? And if you want me to make the changes here, let me know. Thanks for proving this works!

timostrunk commented:
Thanks for the review and sorry for the bogus build.sh change, which I reverted. I added the ignore to the package and agree with your assessment.
Edit: I also edited the Windows cmake variables.

loriab commented:
@conda-forge-admin, please rerender

loriab commented:
Thanks for the changes @timostrunk , it lgtm!

I missed that you had rerendered locally, so I did it here in hopes that the windows fail was a fluke.

timostrunk commented:
Thanks, the windows build error message really looks like a fluke though.

timostrunk commented:
I must have misunderstood the ignore_run_exports, because OSX is also failing:
```
  The Fortran compiler

    "/Users/runner/miniforge3/conda-bld/libint_1704224848023/_build_env/bin/arm64-apple-darwin20.0.0-gfortran"

  is not able to compile a simple test program.
ld: library not found for -lgfortran
```

timostrunk commented:
Considering everything is currently failing, I will pop the two feedstock rerenders and check if the builds start.
Edit: didn't help. I will try around a bit soon with the toy example again. For now I just rerendered again

timostrunk commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.28.2, conda-smithy 3.30.2, and conda-forge-pinning 2024.01.02.19.32.12

timostrunk commented:
The windows build issue is known.
![image](https://github.com/conda-forge/libint-feedstock/assets/3316177/3310c9c4-eb6f-4052-8de1-c2e5d4bfee9d)

For OSX-Arm we are most probably running into a bug with ignore_run_exports_from. When using this, something is interpreted as osx64 and something as osx-arm. I tried working around this for the moment by explicitly listing the dependencies to ignore for osx-arm and that appears to be working.

timostrunk commented:
toy_build went through with the configuration pushed now.
My conclusion: Even if we split build explicitly, osx-arm would not work, because ignore_run_exports changes the build environment for arm, which it should not do. I will prepare a minimal example for a bug report on the mamba repo soon. For the moment, I simply disabled building the fortran extension for osx-arm again. Windows resolved itself.

timostrunk commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.28.2, conda-smithy 3.30.2, and conda-forge-pinning 2024.01.03.13.20.35

timostrunk commented:
@loriab I think this would be ready for merge now.

loriab reviewed (approved):
lgtm, thank you!
